File: src/model/worldMap/TrailEdge.java
package model.worldMap;


import core.Logger;
import model.Condition;

@SuppressWarnings("serial")
public class TrailEdge extends MapObject {
	
	/**end point of this edge - destination node*/
	private LocationNode destination;
	/**end point of this edge - where we started*/
	private LocationNode origin;
	/**describes how hard the transit on this trail will be*/
	private int dangerLevel;
	/**length of the trail.  calculated by the distance between the two location nodes.*/
	private float length;
	/**length of the longest trail on map*/
	private static float longestTrail = 0;
	/**length of the shortest trail on map*/
	private static float shortestTrail = 9999;
	/**length of the shortest trail on map*/
	private static float totalTrailLength = 0;
	/**unique id corresponding to this edge.*/
	private final int ID;
	
	
	/**
	 * Make a trail edge that will connect two locationNodes
	 * @param destination
	 * @param origin
	 * @param dangerLevel
	 */
	
	public TrailEdge(LocationNode destination, LocationNode origin, int dangerLevel){
		this.destination = destination;
		this.origin = origin;
		this.ID = TrailEdge.count++;
		this.dangerLevel = dangerLevel;
		this.name ="Trail to " + destination.getName(); // this.getRoughLength() + ", " + this.getDangerRating() + "," + this.getRoughDirection() + 
		this.length = calcDistance(destination.MAP_XPOS, origin.MAP_XPOS, destination.MAP_YPOS, origin.MAP_YPOS);
		if (this.length > TrailEdge.longestTrail){
			TrailEdge.longestTrail = this.length;
		}
		if (this.length < TrailEdge.shortestTrail){
			TrailEdge.shortestTrail = this.length;
		}
		this.quality = new Condition((int) this.length);
		TrailEdge.totalTrailLength += this.length;
		this.visible = false;
	}

	private float calcDistance(double destX, double origX, double destY, double origY){
		float result = 0;
		//capture any spurious math errors or overflow in coords
		try{
			result = (float) Math.sqrt((Math.pow((origX - destX), 2) + Math.pow((origY - destY), 2)));
		}
		catch (Exception e){
			Logger.log("Error calculating distance between two nodes in map generation - bad math in calcDistance", Logger.Level.ERROR);
			System.out.println("bad math in edge distance calculation for edge : " + this.ID + " " + this.toString());
		}
		
		return result;
	}

	@Override
	public String toString(){
		return this.getRoughDirection() + " " + this.name + " from " + this.origin;
	}
	/**
	 * move along a trail a certain distance
	 * @param movement the amount to move along the trail
	 */
	public void advance(double movement) {
		this.quality.decrease(movement);
	}
	
	/**
	 * Returns the amount of distance left to go for this trail.
	 * @return Distance left to travel
	 */
	public double getDistanceToGo() {
		return quality.getCurrent();
	}
	
	/**
	 * Returns a verbal approximation of distance left to go on the trail
	 * @return An approximation of the distance left.
	 */
	public String getRoughDistanceToGo() {
		double current = 1 - quality.getPercentage();
		String str;
		if (current < .05) {
			str = "Just starting toward ";
		} else if (current < .25) {
			str = "Nowhere close to ";
		} else if (current < .5) {
			str = "Getting closer to ";
		} else if (current < .75) {
			str = "More than halfway to ";
		} else if (current < .95) {
			str = "Just a little further to ";
		} else {
			str = "On the outskirts of ";
		}
		return str;
	}

	/**
	 * Returns a verbal approximation of the direction of the trail, based on the y coords of the destination and the origin
	 * @return A string representation of the direction of the trail.
	 */
	public String getRoughDirection(){
		String resStr;
		
		if (this.destination.MAP_YPOS > this.origin.MAP_YPOS ) {//northwestern trail
			resStr =  "North West";
		} else if (this.destination.MAP_YPOS < this.origin.MAP_YPOS ) {
			resStr =  "South West";		
		} else {
			resStr =  "West";		
		}
		//if the rank between destinations is the same, make only s or n, so return only first char
		if (this.destination.getRank() == this.origin.getRank()) {
			if (resStr.indexOf(' ') > 0) {
				resStr = resStr.substring(0, resStr.indexOf(' '));
			}
		}	
		return resStr;
		
	}
	/**
	 * return the destination for this trail
	 * @return the destination node
	 */
	public LocationNode getDestination(){
		return this.destination;
	}
	/**
	 * return the origin of this trail
	 * @return the origin node
	 */
	public LocationNode getOrigin(){
		return this.origin;
	}
	
	/**
	 * return the danger level of this trail
	 * @return The danger level of this trail
	 */
	public int getDangerLevel () {
		return this.dangerLevel;
	}
	
	/**
	 * Return the danger level of the trail in terms of how travelled it is
	 *  - established and well travelled trails are very easy, wilderness and indian lands trails are very difficult
	 */
	public String getDangerRating () {
		String str;
		if (dangerLevel < 10) {
			str = "Established";
		} else if (dangerLevel < 40) {
			str = "Well-Travelled";
		} else if (dangerLevel < 55) {
			str = "Rarely Travelled";
		} else if (dangerLevel < 90) {
			str = "Wilderness";
		} else {
			str = "Indian Lands";
		}
		return str;
	}
	
	public static void resetTrails(){
		TrailEdge.longestTrail = 0;
		TrailEdge.shortestTrail = 9999;
		TrailEdge.totalTrailLength = 0;	
	}

	@Override
	public String debugToString() {
		String retVal = "";
		retVal += "\n\tLength : " + this.length +  " = " + this.getRoughLength() +  " Danger : " + this.dangerLevel +  " = \"" + this.getDangerRating() + "\" Trail ID : " + this.ID + " Trail Name : \"" + this.name + "\" Rank : " + this.origin.getRank() + " to " + this.destination.getRank() + "\n";
		//retVal += "\tTrail from " + this.origin.getName() + " to " + this.destination.toString() + "that is " + this.length + " miles long\n";
		return retVal;
	}

	/**
	 * Gives a verbal approximation of the trail length relative to the other trails on the map
	 * magic numbers are bad, m'kay?
	 * @return An approximation of the trail length.
	 */
	public String getRoughLength() {
		String str;
		if (length < (.1 * (TrailEdge.totalTrailLength/TrailEdge.count))) {
			str = "Very Short";
		} else if (length < (.5 * (TrailEdge.totalTrailLength/TrailEdge.count))) {
			str = "Short";
		} else if (length <  (TrailEdge.totalTrailLength/TrailEdge.count)) {
			str = "Average";
		} else if (length < (1.5 * (TrailEdge.totalTrailLength/TrailEdge.count))) {
			str = "Long";
		} else if (length < (2 * (TrailEdge.totalTrailLength/TrailEdge.count))) {
			str = "Very Long";
		} else {
			str = "Endless";
		}
		return str;
	}
	
}// class TrailEdge

File: web/src/model/TrailEdge.ts
import { MapObject } from "./MapObject";
import { LocationNode } from "./LocationNode";
import { Condition } from "./Condition";

export class TrailEdge extends MapObject {
  private destination: LocationNode;
  private origin: LocationNode;
  private dangerLevel: number;
  private length: number;
  private static longestTrail: number = 0;
  private static shortestTrail: number = 9999;
  private static totalTrailLength: number = 0;
  private readonly ID: number;
  private taken: boolean = false;

  constructor(destination: LocationNode, origin: LocationNode, dangerLevel: number) {
    super();
    this.destination = destination;
    this.origin = origin;
    this.ID = MapObject.count++;
    this.dangerLevel = dangerLevel;
    this.name = "Trail to " + destination.getName();
    this.length = this.calcDistance(
      destination.MAP_XPOS,
      origin.MAP_XPOS,
      destination.MAP_YPOS,
      origin.MAP_YPOS
    );

    if (this.length > TrailEdge.longestTrail) {
      TrailEdge.longestTrail = this.length;
    }
    if (this.length < TrailEdge.shortestTrail) {
      TrailEdge.shortestTrail = this.length;
    }

    this.quality = new Condition(0, this.length);
    TrailEdge.totalTrailLength += this.length;
    this.visible = false;
  }

  private calcDistance(destX: number, origX: number, destY: number, origY: number): number {
    try {
      return Math.sqrt(Math.pow(origX - destX, 2) + Math.pow(origY - destY, 2));
    } catch (e) {
      console.error("Error calculating distance between two nodes in map generation", e);
      return 0;
    }
  }

  advance(movement: number): void {
    this.quality.decrease(movement);
  }

  getDistanceToGo(): number {
    return this.quality.getCurrent();
  }

  getRoughDistanceToGo(): string {
    const current = 1 - this.quality.getPercentage();
    if (current < 0.05) {
      return "Just starting toward ";
    } else if (current < 0.25) {
      return "Nowhere close to ";
    } else if (current < 0.5) {
      return "Getting closer to ";
    } else if (current < 0.75) {
      return "More than halfway to ";
    } else if (current < 0.95) {
      return "Just a little further to ";
    } else {
      return "On the outskirts of ";
    }
  }

  getRoughDirection(): string {
    let resStr: string;

    if (this.destination.MAP_YPOS > this.origin.MAP_YPOS) {
      resStr = "North West";
    } else if (this.destination.MAP_YPOS < this.origin.MAP_YPOS) {
      resStr = "South West";
    } else {
      resStr = "West";
    }

    if (this.destination.getRank() === this.origin.getRank()) {
      const spaceIndex = resStr.indexOf(' ');
      if (spaceIndex > 0) {
        resStr = resStr.substring(0, spaceIndex);
      }
    }

    return resStr;
  }

  getDestination(): LocationNode {
    return this.destination;
  }

  getOrigin(): LocationNode {
    return this.origin;
  }

  getDangerLevel(): number {
    return this.dangerLevel;
  }

  getDangerRating(): string {
    if (this.dangerLevel < 10) {
      return "Established";
    } else if (this.dangerLevel < 40) {
      return "Well-Travelled";
    } else if (this.dangerLevel < 55) {
      return "Rarely Travelled";
    } else if (this.dangerLevel < 90) {
      return "Wilderness";
    } else {
      return "Indian Lands";
    }
  }

  getRoughLength(): string {
    const avgLength = TrailEdge.totalTrailLength / MapObject.count;
    if (this.length < 0.1 * avgLength) {
      return "Very Short";
    } else if (this.length < 0.5 * avgLength) {
      return "Short";
    } else if (this.length < avgLength) {
      return "Average";
    } else if (this.length < 1.5 * avgLength) {
      return "Long";
    } else if (this.length < 2 * avgLength) {
      return "Very Long";
    } else {
      return "Endless";
    }
  }

  getCurrTrailLocationX(): number {
    const percentComplete = 1 - this.quality.getPercentage();
    return this.origin.getPlayerMapX() + 
           (this.destination.getPlayerMapX() - this.origin.getPlayerMapX()) * percentComplete;
  }

  getCurrTrailLocationY(): number {
    const percentComplete = 1 - this.quality.getPercentage();
    return this.origin.getPlayerMapY() + 
           (this.destination.getPlayerMapY() - this.origin.getPlayerMapY()) * percentComplete;
  }

  isTaken(): boolean {
    return this.taken;
  }

  setTaken(taken: boolean): void {
    this.taken = taken;
  }

  static resetTrails(): void {
    TrailEdge.longestTrail = 0;
    TrailEdge.shortestTrail = 9999;
    TrailEdge.totalTrailLength = 0;
  }

  debugToString(): string {
    return `\n\tLength: ${this.length} = ${this.getRoughLength()} Danger: ${this.dangerLevel} = "${this.getDangerRating()}" Trail ID: ${this.ID} Trail Name: "${this.name}" Rank: ${this.origin.getRank()} to ${this.destination.getRank()}\n`;
  }

  toString(): string {
    return `${this.getRoughDirection()} ${this.name} from ${this.origin}`;
  }
}
